src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -0,0 +1,256 @@
+                Debug.Assert(encodedValue != null);
+                var reader = new CborReader(encodedValue.Value);
+
+                if (label.LabelAsInt32 != null)
+                {
+                    Debug.Assert(label.LabelAsString == null);
+                    // all known headers are integers.
+                    ValidateKnownHeaderValue(label.LabelAsInt32.Value, reader.PeekState(), reader);
+                }
+                else
+                {
+                    Debug.Assert(label.LabelAsString != null);
+                    reader.SkipValue();
+                }
+
+                if (reader.BytesRemaining != 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                }
+            }
+
+            static void ValidateKnownHeaderValue(int label, CborReaderState? initialState, CborReader? reader)
+            {
+                switch (label)
+                {
+                    case KnownHeaders.Alg:
+                        if (initialState != CborReaderState.NegativeInteger &&
+                            initialState != CborReaderState.UnsignedInteger &&
+                            initialState != CborReaderState.TextString)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        }
+                        reader?.SkipValue();
+                        break;
+                    case KnownHeaders.Crit:
+                        throw new NotSupportedException(); // TODO
+                        //if (reader == null)
+                        //{
+                        //    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        //}
+
+                        //int? length = reader.ReadStartArray();
+                        //if (length.GetValueOrDefault() < 1)
+                        //{
+                        //    throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        //}
+
+                        //for (int i = 0; i < length; i++)
+                        //{
+                        //    CborReaderState state = reader.PeekState();
+                        //    if (state != CborReaderState.NegativeInteger &&
+                        //        state != CborReaderState.UnsignedInteger &&
+                        //        state != CborReaderState.TextString)
+                        //    {
+                        //        throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapHeaderDoesNotAcceptSpecifiedValue, label));
+                        //    }
+                        //    reader.SkipValue();
+                        //}
+                        //reader.ReadEndArray();
+                        //break;
```

Comment:
What's the plan for all this commented out code / TODOs?  That's to be addressed in this PR, or subsequently?

Reply:
Subsequent PRs will address this, I can remove the comment-out code and bring it back when it is time.